File: totvs-camera-vision/src/main/java/com/totvs/camera/vision/stream/BroadcastVisionStream.kt
package com.totvs.camera.vision.stream

import androidx.annotation.GuardedBy
import com.totvs.camera.vision.VisionObject

/**
 * [VisionStream] that keep a reference of all the receivers that have connected
 * to it and dispose then appropriately.
 */
internal class BroadcastVisionStream : VisionStream<VisionObject> {

    @GuardedBy("this")
    private val receivers = mutableListOf<VisionReceiver<VisionObject>>()

    /**
     * Broadcast to receivers the incoming [VisionObject]
     */
    @Synchronized
    fun broadcast(entity: VisionObject) {
        receivers.forEach { it.send(entity) }
    }

    /**
     * Add a [receiver] to this stream
     */
    @Synchronized
    fun add(receiver: VisionReceiver<VisionObject>) {
        receivers.remove(receiver)
        receivers.add(receiver)
    }

    /**
     * Remove [receiver] from this stream
     */
    @Synchronized
    fun remove(receiver: VisionReceiver<VisionObject>) {
        receivers.remove(receiver)
    }

    override fun connect(receiver: VisionReceiver<VisionObject>): Connection {
        return ReceiverConnection(receiver.also {
            add(receiver)
        })
    }

    /**
     * Internal class that automatically remove the receiver from
     * the registry of receiver at [disconnect] moment.
     */
    inner class ReceiverConnection(
        private val receiver: VisionReceiver<VisionObject>
    ) : Connection {
        override fun disconnect() = remove(receiver)
    }
}

File: ios/TOTVSCameraKit/Vision/Stream/BroadcastVisionStream.swift
//
//  BroadcastVisionStream.swift
//  TOTVSCameraKit
//
//  Created by Jansel Rodriguez on 7/11/20.
//  Copyright © 2020 Jansel. All rights reserved.
//

import Foundation

class BroadcastVisionStream : VisionStream<VisionObject> {
    
    private var receivers = [VisionReceiver<VisionObject>]()
    
    func broadcast(value: VisionObject) {
        receivers.forEach { r in r.send(value: value) }
    }
    
    func add(receiver: VisionReceiver<VisionObject>) {
        receivers.removeAll(where: { $0 === receiver })
        receivers.append(receiver)
    }
    
    func remove(receiver: VisionReceiver<VisionObject>) {
        receivers.removeAll(where: { $0 === receiver })
    }
    
    override func connect(receiver: VisionReceiver<VisionObject>) -> Connection {
        add(receiver: receiver)
        return ReceiverConnection(owner: self, receiver: receiver)
    }
    
    private class ReceiverConnection : Connection {
        weak var owner: BroadcastVisionStream? = nil
        
        let receiver: VisionReceiver<VisionObject>
        
        init(owner: BroadcastVisionStream, receiver: VisionReceiver<VisionObject>) {
            self.owner = owner
            self.receiver = receiver
        }
        
        func dicsonnect() {
            owner?.remove(receiver: receiver)
        }
    }
}
